tests/error/comprehension_errors/guarded3.err

```diff
@@ -0,0 +1,7 @@
+Guppy compilation failed. Error in file $FILE:24
+
+22:    @guppy(module)
+23:    def foo(qs: linst[MyStruct]) -> linst[qubit]:
+24:        return [s.q2 for s in qs if bar(s.q1)]
+                            ^
+GuppyTypeError: Field `s.q2` with linear type `qubit` is not used on all control-flow paths of the list comprehension
```

Comment:
I think talking about "control-flow paths" of a list comprehension could be a bit misleading, not sure what the better wording is. I'm thinking `Field ... with linear type ... may not be used by this comprehension` but that's not great either

Reply:
Ideally, you'd want to say that `s.q2` is not used if `bar(s.q1)` is false.

Our dataflow analysis can already exactly identify which branch leads to a non-use, so we are actually not far off from having this better error message. However, we need to extend our error framework so that it can point to this location in the message.

I think something like this would be nice:

```
23:    def foo(qs: linst[MyStruct]) -> linst[qubit]:
24:        return [s.q2 for s in qs if bar(s.q1)]
                            ^ Field `s.q2` with linear type `qubit` is not always used

24:        return [s.q2 for s in qs if bar(s.q1)]
                                       --------- Field is not used if this expression is `False`
```

Reply:
Added to issue #306